Copybutton

s-light opened the pull request:
closes #182 

i just added buttons at different points..
the image copy buttons are not working yet..
(was just an experiment)

b-g commented:
Hi @s-light, a belated happy new year + sorry for the slow response! Needed a break. 

Many thanks for this! I think I find all 3 positions not ideal ... what about next to Unicode?

![Untitled](https://github.com/hfg-gmuend/openmoji.org/assets/480224/f2bb7c85-1eb5-4ac9-b70f-2bcfb4952468)

s-light commented:
iam fine with it..
so i will move it and let you know.

s-light commented:
orig:
![image](https://github.com/hfg-gmuend/openmoji.org/assets/1340319/b28508d5-ce80-45e0-bdf5-b523fc832a28)
moved to new location:
![image](https://github.com/hfg-gmuend/openmoji.org/assets/1340319/24efefc9-ad40-4cdb-ae43-7a147745bd0d)
this is the position it goes without extra styling when in the same table cell  - but not inside the heading2:
```astro
        <tr>
          <td>
            <h2>Unicode</h2>
            <ButtonCopy content={generalUtil.getEmojiCombinationString(openmoji.hexcode)}></ButtonCopy>
          </td>
          <td>
            <a target="_blank" class="redlink" href={urlUnicode}>
              {openmoji.hexcode}
            </a>
          </td>
        </tr>
```


is this ok for you?

b-g commented:
Hi @s-light, 👍! I guess this is the easiest way for positioning the button. 

If you want to have the image as well ... maybe a hack could be to add in addition to the button also a listener to `CMD/CTRL + C` for coping the image to the clipboard?

b-g commented:
OK 👍. Makes sense. Please let me know once it is final. Happy to merge and test. Cheers!

s-light commented on public/js-code/copy-clipboard.js:
i first tried to add the base *copy* action thing as a new *utility* js file.

s-light commented on src/components/shared/ButtonCopy/ButtonCopy.astro:
and created a new astro component.

b-g commented:
@s-light Danke! Sieht grundsätzlich gut aus. Spontan fallen zwei Sachen auf (hab Stephan Bogner für Feedback gefragt):

* „global“ sollte nicht nötig sein im SCSS da ja auf die Komponente beschränkt

* Die JS-Logik ist etwas "komisch" eingebunden. Die Import-Logik die man aus React kennt gibt es nur bedingt in Vanilla Astro. 
  * Empfehlung für „old school“-approach (à la jQuery) einmal die Funktion zu deklarieren und dann in der Komponente nur auszuführen.
   *  Beispiel dafür wäre „<script src="/js-code/library-navigation.js"/>“ [hier](https://github.com/hfg-gmuend/openmoji.org/blob/master/src/pages/library/index.astro). (In dem Fall werden sogar die Click Handlers in der JS-Datei definiert … ist nicht super ideal, aber da wir Abhängigkeiten zu jQuery nicht easy zu vermeiden)

s-light commented:
the copy button for the unicode working.

added Ctrl+Click functionality to SVG + PNG download buttons.

the first SVG button gets messed up by something else - 
there are some not visible - maybe not closed - a tags...
i have no more time to dig into this..

SVG is not copyable in firefox for now.
(firefox does currently not allow the MIME-type image/svg+xml)
png is working fine.

i think it is fine for now.
good enough for now - save enough to try ;-)

s-light commented:
@b-g fehlt dir aktuell noch irgendwas spezifisches?

b-g commented:
Danke dafür! Ja, dein PR hat 2430 Files changes :) Kannst du den bitte aufräumen, so möchte ich den nicht mergen da nicht zu überschauen was du alles geändert hast. Sorry, ich hoffe du verstehst. Aber klar ... wir hätten Copy Button gerne auf OpenMoji.org

s-light commented:
soory...
hatte ein build ausgeführt (das landet ja in docs)  und aus versehen auch mit committed..

eventuell könnte man da eine pre-commit warnung oder sowas ausgeben?!
(nur so ne idee - keine ahnung ob / wie sich das mit git umsetzen liese)

s-light commented:
i merged your fixes.

now #184 is also fixed for me.

while testing i encounterd a nother unexpected behavior:

i opened http://127.0.0.1:3000/library/emoji-1F600
and then tried to go the emoji befor (with the link down left)
http://127.0.0.1:3000/library/emoji-2BEB
and got a 
```
500: Internal Error

Cannot read property 'description' of undefined

TypeError: Cannot read property 'description' of undefined
    at eval (/src/components/library/EmojiDetail/EmojiDetail.astro:30:39)
    at renderToString (../../../src/runtime/server/index.ts:408:42)
    at Module.renderComponent (../../../src/runtime/server/index.ts:141:38)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
```
your quick fix could not handle if there is no emojipedia entry..
so i added
```js
if (emojiPedia == undefined) {
  emojiPedia = {
    description: ""
  };
}
```
hopefully this is enough for now ;-)

b-g commented:
Hallo @s-light, erstmal ein dickes Sorry, dass du mit in den fiesen emojiPedia.description Bug reingezogen worden bist. Sorry! Kam nicht von uns sondern mit buggy Daten von EmojiPedia. Ich habe jetzt an der Wurzel des Problems angesetzt und es direkt beim Datenimport gelöst. https://github.com/hfg-gmuend/openmoji.org/commit/2f689469cc1017897aabdb9d03f382a0acef1092 und auch die Quickfixes aufgeräumt ... jetzt klappt es also wieder. 

Dabei ist mir allerdings aufgefallen:
- die Komponente von dir wirft Fehler ...
- weiter unten ist noch ein TODO ...

Kannst du beides auch noch aufräumen, damit wir keine neue Probleme erben? :)

Danke!

<img width="1500" alt="Screen Shot 2024-03-06 at 09 43 59" src="https://github.com/hfg-gmuend/openmoji.org/assets/480224/c29ba4f6-96c6-4ea7-82f1-739a07bdadbb">

<img width="1366" alt="Screen Shot 2024-03-06 at 09 50 55" src="https://github.com/hfg-gmuend/openmoji.org/assets/480224/866f5853-aabc-46a4-97c0-7c1c14659120">

s-light commented:
danke fürs aufräumen ;-)
klar räum ich da noch auf!

- das ToDo ist allerdings nicht von mir!  das war schon drin!

mir scheint die fehler sind jetzt allerdigns irgendwie neu..
und mich wundert das das beispiel wie man das macht aus der astro docu die gleichen fehler wirft..
https://docs.astro.build/en/guides/client-side-scripts/#using-script-in-astro

>   By default, <script> tags are processed by Astro.
>
>    Any imports will be bundled, allowing you to import local files or Node modules.
>    The processed script will be injected into your page’s <head> with [type="module"](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules).
>    TypeScript is fully supported, including importing TypeScript files.
>    If your component is used several times on a page, the script will only be included once.

ich hab einfach zum testen meinen code raus genommen und nur basierend auf dem beispiel ein minimal gemacht:
https://docs.astro.build/en/guides/client-side-scripts/#pass-frontmatter-variables-to-scripts
```js
class ButtonDownloadImage extends HTMLElement {
        constructor() {
            super();
            console.log("ButtonDownloadImage constructor", this);
        }
        connectedCallback() {
            // buttonConnectedCallback(this);
        }
    }

    customElements.define("button-download-image", ButtonDownloadImage);
</script>
```

da kommt der gleiche fehler typ:
```
09:55:19.284 ButtonDownloadImage constructor 
<button-download-image class="astro-YKYNXPO3">
emoji-1F334:334:21
09:55:19.285 ButtonDownloadImage constructor 
<button-download-image>
emoji-1F334:334:21
09:55:19.285 Uncaught SyntaxError: redeclaration of let ButtonDownloadImage
    <anonymous> http://127.0.0.1:3000/library/emoji-1F334:357
emoji-1F334:357:9
09:55:19.285 ButtonDownloadImage constructor 
<button-download-image>
emoji-1F334:334:21
09:55:19.285 Uncaught SyntaxError: redeclaration of let ButtonDownloadImage
    <anonymous> http://127.0.0.1:3000/library/emoji-1F334:386
emoji-1F334:386:9
09:55:19.285 ButtonDownloadImage constructor 
<button-download-image>
emoji-1F334:334:21
09:55:19.286 Uncaught SyntaxError: redeclaration of let ButtonDownloadImage
    <anonymous> http://127.0.0.1:3000/library/emoji-1F334:413
emoji-1F334:413:9
09:55:19.288 [vite] connecting... client.ts:22:8
09:55:19.333 [vite] connected. client.ts:58:14
``` 

das heißt genau das bundling und inkludieren funktioniert nicht.
liegt das an der astro version die noch verwendet wird?

s-light commented:
hab weiter gegrabben....
und die docu für v0.23 gefunden:
https://web.archive.org/web/20220521182404/https://docs.astro.build/en/core-concepts/astro-components/#client-side-scripts

und da steht das selbe - 
scripts sollten vor verarbeitet werden..

habe nun ein workaround eingebaut der prüft ob die componente schon vorhanden ist..
das passt nun.

b-g commented:
@s-light Many thanks!! Merged! Will deploy in the next days.
